### PR TITLE
fix(goto): send UA Client Hints consistent with the user agent

### PR DESCRIPTION
**Pull request by Kikobeats**
## What changed

`goto` overrides the user agent with `page.setUserAgent(string)`. That call wipes UA Client Hints, so every navigation with a Chrome user agent sent no `sec-ch-ua*` headers, and `navigator.userAgentData` reported empty brands and platform while `navigator.userAgent` claimed Chrome. Pages that read Client Hints saw a browser that doesn't exist.

`goto` now calls `page.setUserAgent({ userAgent, userAgentMetadata, platform })`, with the metadata derived from the user agent string (`packages/goto/src/client-hints.js`):

- **Brands:** Chromium's GREASE algorithm (seeded by the major version). Checked against real Chrome for Testing 152, Google Chrome 152 and Google Chrome Canary 154: brand names, versions and order match exactly.
- **fullVersionList / fullVersion:** the user agent's version. A reduced `x.0.0.0` user agent takes the browser's full version when the majors match. The version comes from the existing per-browser cache (`chromeVersionFromBrowser`). The lookup is bounded to half the action timeout, and its result is cached per browser whether it is the browser version or the fallback after a failed or hung `browser.version()`. Only the first goto on a browser pays for it; when the lookup doesn't settle, the override goes out with the user agent's own version.
- **platform / platformVersion / architecture / bitness / model / mobile / formFactors**, following the user agent's OS:

| User agent | `navigator.platform` | platform | platformVersion | architecture / bitness | mobile |
|---|---|---|---|---|---|
| macOS (`Macintosh`) | `MacIntel` | macOS | 26.6.2 (measured on real Chrome 152) | arm / 64 | false |
| Windows NT 10.0 (`Win64; x64`, `WOW64`) | `Win32` | Windows | 19.0.0 | x86 / 64 (WOW64: 32, wow64 true) | false |
| Linux (`X11; Linux x86_64`) | `Linux x86_64` | Linux | `""` (as Chromium reports on Linux) | x86 / 64 | false |
| Chrome OS (`CrOS`) | `Linux <arch>` | Chrome OS | from the user agent | from the user agent | false |
| Android (`Mobile` phone, tablet) | `Linux armv8l` | Android | from the user agent | empty | true / false, model from the user agent |
| Android WebView (`; wv)`) | `Linux armv8l` | Android | from the user agent | empty | brand `Android WebView` instead of `Google Chrome` |

Left untouched, keeping the previous behavior with no invented brands:
- Chromium-based browsers with their own product token (Samsung Internet, Opera, Yandex, Edge, Vivaldi)
- Non-Chromium user agents (Safari/iPhone, Firefox, `googlebot`, Googlebot smartphone)
- HeadlessChrome

## Before / after

Measured on Chrome 152 with a local server that reads the request headers, plus the values the page itself reads.

| Signal | master | this PR |
|---|---|---|
| `sec-ch-ua*` request headers (default device) | 0 | 6 (`sec-ch-ua`, `-mobile`, `-platform`, `-platform-version`, `-full-version-list` and a high-entropy hint requested via Accept-CH) |
| `navigator.userAgentData.brands` (default device) | `[]` | Chromium/152, Not?A_Brand/24, Google Chrome/152 |
| `getHighEntropyValues().platform` | `""` | `macOS` |
| Galaxy S8 `navigator.platform` | `MacIntel` (host value leaking next to an Android user agent) | `Linux armv8l` |
| Galaxy S8 `sec-ch-ua-mobile` | absent | `?1` |
| iPhone 15 (Safari user agent) | no hints | no hints (unchanged) |

## How it was tested

- `test/unit/client-hints.js`:
  - GREASE order for Chrome 152 and 154
  - reduced user agent full version
  - macOS, Windows, Android, Android WebView, Linux (empty platformVersion) and Chrome OS mapping
  - Samsung Internet, Opera, Yandex, Edge and Vivaldi excluded
  - non-Chromium user agents excluded
- `test/unit/client-hints.js` also drives `goto` with a mock page:
  - a `browser.version()` that never settles: goto still navigates and sends the metadata (fails on `e02a61b2`, which awaited the lookup unbounded)
  - a reduced user agent header on the first navigation gets the browser full version `152.0.7977.83` (fails on `b193d281`, which only read a cold cache and sent `152.0.0.0`)
  - a hung or failed lookup runs once across two gotos on the same browser (fails on `fc008fb8`, which retried and waited on every goto)
  - Android maps to `navigator.platform` exactly `Linux armv8l`, as real Chrome on Android reports
- `test/index.js` (real Chrome, local server):
  - default Chrome user agent sends `sec-ch-ua` and brands matching its major, `platform: macOS`
  - Windows and Linux user agents send matching `sec-ch-ua-platform` and `navigator.platform`, Linux platformVersion `""`
  - Galaxy S8 reports Android and `Linux armv8l`
  - iPhone and `googlebot` get no brands
- With master's `src/index.js` and no `client-hints.js`, the unit file can't load and the Chrome, Windows/Linux and Android e2e tests fail. The non-Chromium guard test passes on both, by design.
- Full `@browserless/goto` suite and `standard` lint.

## Follow-ups

- **Hardcoded platform values:** macOS platformVersion `26.6.2` with architecture `arm` (measured on a real Apple Silicon Chrome 152) and Windows platformVersion `19.0.0` (a fixed table value, not measured) are constants, not derived from the host or the user agent.
- **Other products:** Edge, Opera and Samsung Internet (also Yandex and Vivaldi) get no hints, same as master, rather than guessed brands.
- **Default-device lookup:** master's default-device path awaits `chromeVersionFromBrowser(page)` with no timeout, so a hung `browser.version()` there still blocks goto when nothing cached a result first. Unchanged here.
- **Unverified on real devices:** Linux `""` and the Android WebView brand follow Chromium's source, not a real Linux or Android device.
- **Screen:** emulating a screen consistent with the device viewport is a separate PR (#923).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_0116zka7w2WKPSi3kqHyVAET

---

> [!NOTE]
> **Medium Risk**
> Changes navigation fingerprinting for all Chrome UA overrides; incorrect hint synthesis could break sites that gate on Client Hints, though scope is limited to recognized Chrome UAs with broad test coverage.
> 
> **Overview**
> **`goto` no longer wipes UA Client Hints when overriding the user agent.** Instead of `page.setUserAgent(string)`, it now passes `userAgentMetadata` and `platform` derived from the Chrome user agent string via new `getClientHints` in `client-hints.js`.
> 
> For **Google Chrome–style** user agents, hints include GREASE brand ordering, `fullVersion` / `fullVersionList`, and OS-specific fields (macOS, Windows, Linux, Chrome OS, Android, Android WebView). Reduced `x.0.0.0` UAs can pick up the real browser build from a **timeout-bounded** `browser.version()` lookup, cached per browser so hung lookups do not block navigation. **Non-Chrome** UAs (Safari, Firefox, bots, HeadlessChrome, Edge/Opera/Samsung, etc.) still get **no** fabricated hints.
> 
> Integration and unit tests cover `sec-ch-ua*` headers, `navigator.userAgentData`, platform mapping (including Galaxy S8 → `Linux armv8l`), and version-lookup edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9234806452f83b88bfa03f7b9c93fa9b0f33bc9a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added support for generating accurate Chrome User-Agent Client Hints across desktop, mobile, tablet, and WebView environments.
  * Improved platform, browser version, architecture, device, and mobile status detection for supported browsers.
  * Custom user-agent overrides now include corresponding client hints and full browser versions when available.

* **Bug Fixes**
  * Prevented delayed or failed browser version lookups from blocking navigation or user-agent configuration.
  * Improved handling of reduced user-agent versions and unsupported browser formats.

### CODE REVIEW

File: packages/goto/src/index.js [hunk 1]
```diff
@@ -432,9 +433,10 @@ module.exports = ({ defaultDevice = 'Macbook Pro 13', timeout: globalTimeout, ..
       }
 
       if (userAgent) {
+        const clientHints = getClientHints(userAgent, await chromeVersionFromBrowser(page))
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/goto/src/index.js]: Fixed in 9f29803b. The client hints no longer await chromeVersionFromBrowser: goto reads the version from the per-browser cache that the default-device path already fills, and passes undefined when it isn't cached. In that case fullVersion falls back to the user agent's version. The new unit test 'a browser version lookup that never settles does not block the user agent override' uses a browser.version() that never resolves. It fails on e02a61b2 (goto stays blocked) and passes now, with setUserAgent still receiving the metadata. The awaited lookup for device user agents is unchanged from master.

[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/goto/src/index.js]: Correction: the fix landed as b193d281, not 9f29803b; I amended that commit before pushing. The version is now read inline with chromeVersionCache.get(page.browser?.()), same behavior as described above. That also keeps this PR free of merge conflicts with #923 in either order.

File: packages/goto/src/index.js [hunk 2]
```diff
@@ -432,9 +433,10 @@ module.exports = ({ defaultDevice = 'Macbook Pro 13', timeout: globalTimeout, ..
       }
 
       if (userAgent) {
+        const clientHints = getClientHints(userAgent, chromeVersionCache.get(page.browser?.()))
         prePromises.push(
           run({
-            fn: page.setUserAgent(userAgent),
+            fn: page.setUserAgent({ userAgent, ...clientHints }),
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/goto/src/index.js]: Fixed in fc008fb8. `goto` now looks up the browser version with `chromeVersionFromBrowser`, bounded to half the action timeout, before building the hints, so a reduced user agent header gets the full version on the first navigation. If the lookup does not settle, the override still goes out with the user agent version. Regression test "a reduced user agent header gets the browser full version on the first navigation" in `test/unit/client-hints.js` fails on b193d281 (`152.0.0.0` vs `152.0.7977.83`) and passes here.

**Comment by Kikobeats** [comment]
@coderabbitai review
